### PR TITLE
WIP: Example RSA with spiking data

### PR DESCRIPTION
**Pull request by kingjr**
Here is an example that aims at providing 3 small enhancements:
1) introduce spiking data to the package
2) introduce a dataset with a high number of classes
3) have an example of RSA

I thus made the example with some data from the Desimone's lab: http://www.readout.info/downloads/datasets/zhang-desimone-7-object-dataset/. It's typical of monkey electrophy where they record different neurons at different moments. (This means that multivariate analyses won't be particularly more powerful than univariate ones, because there is no noise covariance across most of the channels.)

The size of the dataset is small. However, I'm not sure how to add it to the rep. If you guys agree to add it, could someone help me adding it to the package in a clean way?

With regard to the RSA
- I don't think we need to create a special class or function; it works with pure sklearn
- We could do easily improve the method by making probabilistic predictions, logarithmic scoring etc, but I'm keeping it to its classic form.
- The resulting figures are:

Decoding over time with the new `SearchLight`:
score

Confusion matrix on one time sample:
cm

Dimensionality reduction with Multidimensional Scaling:
mds

### CODE REVIEW

File: examples/decoding/plot_decoding_rsa.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ Science 210.4468 (1980): 390-398.
+[2] Laakso, A. & Cottrell, G.. "Content and cluster analysis:
+ assessing representational similarity in neural systems." Philosophical
+ psychology 13.1 (2000): 47-76.
+[3] Kriegeskorte, N., Marieke, M., & Bandettini.  P. "Representational
+ similarity analysis-connecting the branches of systems neuroscience."
+ Frontiers in systems neuroscience 2 (2008): 4.
+[4] Zhang, Ying, et al. "Object decoding with attention in inferior temporal
+ cortex." Proceedings of the National Academy of Sciences 108.21 (2011):
+ 8850-8855.
```
Comment (wmvanvliet) [inline examples/decoding/plot_decoding_rsa.py]: I find that this paper introduces the topic of RSA very well:

Kriegeskorte, N., Mur, M., & Bandettini, P. a. (2008). Representational similarity analysis - connecting the branches of systems neuroscience. Frontiers in Systems Neuroscience, 2(November), 4. http://doi.org/10.3389/neuro.06.004.2008

File: examples/decoding/plot_decoding_rsa.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
+"""Representational Similarity Analysis is used to perform summary statistics
```
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: you need a headline eg.

Representational Similarity Analysis on spiking data
======================================

File: examples/decoding/plot_decoding_rsa.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+# Preprocess the categories
+le = LabelEncoder()
+y['condition'] = y[['cat', 'position']].apply(lambda x: '/'.join(x), axis=1)
+y_condition = le.fit_transform(y['condition'])
+
+##############################################################################
+# Decode over time to identify when information peaks
+clf = make_pipeline(StandardScaler(), LogisticRegression())
+sl = SearchLight(clf, n_jobs=-1)
```
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: don't use n_jobs=-1 for examples

File: examples/decoding/plot_decoding_rsa.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+files = os.listdir(data_path)
+
+classes = ('car', 'couch', 'face', 'flower', 'guitar', 'hand', 'kiwi')
+positions = ('lower', 'middle', 'upper')
+
+###############################################################################
+
+# FIXME Add this spiking data in MNE samples in a fif format?
+# epoched matlab data available from
+# http://www.readout.info/downloads/datasets/zhang-desimone-7-object-dataset/
```
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: it seems like you need to register to get the data

I doubt we have the permission to redistribute it and make it downloadable
by anyone.

or we could ask the desimone lab if they are ok with it provided we do
something like license agreement as done for brainstorm datasets.

File: examples/decoding/plot_decoding_rsa.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.io import loadmat
+from pandas import DataFrame
+from itertools import product
+
+from sklearn.preprocessing import LabelEncoder
+from sklearn.model_selection import StratifiedKFold, StratifiedShuffleSplit
+from sklearn.pipeline import make_pipeline
+from mne.decoding import SearchLight
```
Comment (jona-sassenhagen) [inline examples/decoding/plot_decoding_rsa.py]: So the only part of MNE this uses is the searchlight? :D

**Comment by jona-sassenhagen** [comment]
Hm ... it seems to me this does two things:

1. show how to handle a particular kind of non-MNE data: single spike counts
2. show RSA

Wouldn't it be more gentle to have a single-purpose example that shows how to use RSA for typical MNE data (MEG/EEG/STC)?

**Comment by jona-sassenhagen** [comment]
Or the other way around, wouldn't this basically be a good scikit-learn main example, if it's possible to emulate a sufficient subset of the MNE searchlight code in a few lines?

File: examples/decoding/plot_decoding_rsa.py (hunk 4 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: ok, I copied you in the email, let's see what they say

File: examples/decoding/plot_decoding_rsa.py (hunk 5 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: yes. We *could* make a function that does it all, but I feel that it's better to show how simple this analysis is, rather than making a do-it-all function.

**Comment by kingjr** [comment]
> Hm ... it seems to me this does two things:
>
>    show how to handle a particular kind of non-MNE data: single spike counts
>    show RSA with the searchlight

> Wouldn't it be more gentle to have a single-purpose example that shows how to use RSA/searchlight for typical MNE data (MEG/EEG/STC)?


True, it's a double aim. Maybe it's better to move this to the tutorial than the gallery/examples, especially as this is not pure mne, WDYT?

> Or the other way around, wouldn't this basically be a good scikit-learn main example, if it's possible to emulate a sufficient subset of the MNE searchlight code in a few lines?

No chance they will accept that. RSA is a neuroimaging taxonomy; the rest of ML would just say 'MDS applied to the confusion matrix', and wouldn't devote a section to this ad-hoc summary metrics.

Overall, I think it would be great to include monkey data. All labs that I see write their own code, but their needs are typically a subset of ours: i.e. filtering, time freq, spectra, stats, decoding, plotting etc.

**Comment by jona-sassenhagen** [comment]
I assume it doesn't help to use one of the MNE classes here? (Since the decoding stack is moving to an np array path)

**Comment by jona-sassenhagen** [comment]
I mean, as an intermediate step to store the spike trains.

**Comment by kingjr** [comment]
> I assume it doesn't help to use one of the MNE classes here? (Since the decoding stack is moving to an np array path)

Not I my opinion. Especially because the confusion matrix isn't considered as a scoring metrics, we wouldn't be able to pipe it in our TimeDecoding or GAT object. These objects would only handle the CV under the hood, but I find this is using a heavy tools (convert to Epochs, applies GAT, retrieve predictions, compute CM + MDS) for something that can be pure sklearn.

**Comment by kingjr** [comment]
> I mean, as an intermediate step to store the spike trains.

I hope we'll be able to store the spikes as a .fif file, and remove the present data conversion from the tutorial/example.

**Comment by jona-sassenhagen** [comment]
>I hope we'll be able to store the spikes as a .fif file, and remove the present data conversion from the tutorial/example.

With that, it'd look a bit more MNE-y.

**Comment by agramfort** [comment]
FYI I have a subject from Cichy NN paper but barely played with it yet.

I am sure he would be happy if we used is data for our example.

might not be as clear as the spiking signals...

**Comment by kingjr** [comment]
> FYI I have a subject from Cichy NN paper but barely played with it yet. 

Yes, we can use RSA for this too, I'll send him an email.

In this case, I would use MEG only, especially because it's truly single trials, whereas electrophy records different neurons/channels on different trials.

That said, I'm still +1 for having a spike dataset, but we could separate the issues.

**Comment by agramfort** [comment]
if he is ok I can share with you his data

**Comment by jona-sassenhagen** [comment]
>That said, I'm still +1 for having a spike dataset, but we could separate the issues.

Don't overrate my comment JR, it was just a thought. Maybe it's just perfect to have this example the way it is.

**Comment by Eric89GXL** [comment]
No comments other than to say, thanks for pursuing this, it could open up a nice complementary user base

**Comment by dengemann** [comment]
I like this. Adding spiking data is a good idea.

**Comment by kingjr** [comment]
The authors of the spiking data only allow us to make a link to their website, which needs a registration pages. So I'll split the problem in 2. For the RSA example, I'll try to apply it to Cichy's nature neuro. For the spiking data, I'll ask around whether there's another dataset available.

**Comment by kingjr** [comment]
Working example with Cichy's data using a single subject. There's is a slight clustering of the face stimuli, but we're facing a strong tradeoff between computing speed and decoding accuracy.  Here's we what we get with a reasonable computing time:


figure_1
figure_1-1


I could introduce a bunch of tricks (e.g. I wrote an efficient parallel auc scorer for example, could use shuffle splitting, cross_val_predict with a transformed estimator that uses predict_proba instead of predict so as to uses its internal n_jobs etc), but I think this sort of goes against the spirit of the gallery examples.

File: examples/decoding/plot_decoding_rsa.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+conds = read_csv(fname)
+triggers = set(conds['trigger'])
+inv_conds = dict()
+for val in triggers:
+    sub = conds.query('trigger == %i' % val)
+    inv_conds[val] = sub['condition'].iloc[0]
+
+# Read MEG data
+fname = op.join(data_path, 'sample_subject_0_tsss_mc.fif')
+raw = Raw(fname)
```
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: Raw -> read_raw_fiff

File: examples/decoding/plot_decoding_rsa.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+fname = op.join(data_path, 'visual_stimuli.csv')
+conds = read_csv(fname)
+triggers = set(conds['trigger'])
+inv_conds = dict()
+for val in triggers:
+    sub = conds.query('trigger == %i' % val)
+    inv_conds[val] = sub['condition'].iloc[0]
+
+# Read MEG data
+fname = op.join(data_path, 'sample_subject_0_tsss_mc.fif')
```
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: you're using only a fraction of the data right?

File: examples/decoding/plot_decoding_rsa.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+epochs = Epochs(raw, events=events, decim=4, baseline=None, picks=picks,
+                tmin=.050, tmax=.800)
+
+##############################################################################
+# Representational Similarity Analysis is a neuroimaging-specific
+# appelation to refer to statistics applied to the confusion matrix.
+
+# Classify on the average ERF
+clf = make_pipeline(StandardScaler(),
+                    LogisticRegression(solver='lbfgs'))
```
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: can you explicit C?

**Comment by agramfort** [comment]
category names should be in middle of section is matrix plot

**Comment by agramfort** [comment]
@kingjr just for my curiosity did you manage to make a nicer figure even with more computation time?

**Comment by jona-sassenhagen** [comment]
It would run faster AND look better if only a subset of categories (say, 3) were included, right?

File: examples/decoding/plot_decoding_rsa.py (hunk 6 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: May I ask why? Raw is more generic no?

File: examples/decoding/plot_decoding_rsa.py (hunk 7 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: no, the whole subject. Taking a subset really makes the results drop.

File: examples/decoding/plot_decoding_rsa.py (hunk 8 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: why? Isn't that outside the scope of this example?

**Comment by kingjr** [comment]
> category names should be in middle of section in matrix plot

I'm opened to an easy way of doing this ...

> It would run faster AND look better if only a subset of categories (say, 3) were included, right?

RSA requires a high number of classes

> @kingjr just for my curiosity did you manage to make a nicer figure even with more computation time?

I didn't push too much, but you can clean the matrix a bit with shuffle splits. (This actually comforts me in thinking that the way RSA is solved isn't right, because we need to solve a *lot* of spatial filters before estimating the similarities. But that's for real science, not mne gallery ;))

**Comment by kingjr** [comment]
BTW, what do we do with the data, @agramfort can you take care of including a new dataset?

File: examples/decoding/plot_decoding_rsa.py (hunk 6 above)
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: Raw only reads fif files. Now we can read many files and I've seen people trying Raw on edf files... We use now everywhere (normally) read_raw_fiff

File: examples/decoding/plot_decoding_rsa.py (hunk 8 above)
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: I like to educational reasons to make explicit model parameters even if we use defaults

File: examples/decoding/plot_decoding_rsa.py (hunk 7 above)
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: are you sure this reads the 4 fif files? I thought they were different runs.

Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: My bad; I assumed that a digit after subject meant subject, so yes, I'm only using this run. Do you want to try with all runs? It's a seems a bit heavy to download no?

File: examples/decoding/plot_decoding_rsa.py (hunk 6 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: ok

Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: (BTW, why can't Raw read edf files?)

Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: because...

signature differs. Edf is not as self contained as fif

File: examples/decoding/plot_decoding_rsa.py (hunk 7 above)
Comment (agramfort) [inline examples/decoding/plot_decoding_rsa.py]: let us first see if it makes the figure more convincing

**Comment by kingjr** [comment]
with all runs, classified on the average topography between .100 and .500
figure_1
figure_2

The data reading becomes significant..

File: examples/decoding/plot_decoding_rsa.py (hunk 6 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: my collaborators get confused by this; I reckon we should make `Raw` have an automatic detection based on whether it's a file/folder + name extension.

File: examples/decoding/plot_decoding_rsa.py (hunk 7 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: I don't find the gain to be huge, see image below. LMTKWYT

File: examples/decoding/plot_decoding_rsa.py (hunk 8 above)
Comment (kingjr) [inline examples/decoding/plot_decoding_rsa.py]: ok

**Comment by kingjr** [comment]
The current example is a bit too heavy IMO, but @agramfort you can give it a try.

I don't think we can easily boost the cm without recruiting multiple cores and a high number of overlapping cv splits to reduce the training error.

Here are the resulting figure
figure_rsa_1
figure_rsa_2

**Comment by Eric89GXL** [comment]
+1 for leaving the example light at the expense of performance -- we don't want it to take forever. You could make notes for changes or comment the slow changes out to tell people how to make it do better. That is of course assuming @agramfort doesn't have some trick up his sleeve

**Comment by kingjr** [comment]
I guess we could also add an `epo-.fif` with all trials from all runs. If it's heavily cropped, it won't take that much space. Anyway, your call @agramfort.

File: examples/decoding/plot_decoding_rsa.py (hunk 6 above)
Comment (teonbrooks) [inline examples/decoding/plot_decoding_rsa.py]: cf. https://github.com/mne-tools/mne-python/pull/1930

**Comment by mmagnuski** [comment]
a tiny remark: the commit history is messed up

**Comment by agramfort** [comment]
I would restrict the learning to the first 3 blocks hence 1 third of the
data
and user is free to uncomment a line to make the full monty.

**Comment by kingjr** [comment]
> I would restrict the learning to the first 3 blocks hence 1 third of the data and user is free to uncomment a line to make the full monty.

ok. Can you take care of adding the dataset to mne?

> a tiny remark: the commit history is messed up

ouch what have I done :/

**Comment by kingjr** [comment]
@agramfort I need the cichy dataset to be shipped with MNE to be able to move on with this PR

**Comment by agramfort** [comment]
try to use figshare. See how it's done in:

https://github.com/mne-tools/mne-python/pull/3205/files#diff-751494890378ef87b64cacb16d553db0

**Comment by kingjr** [comment]
@jaeilepp would you be interested in finishing the PR. It's mainly about adding the dataset to S3

**Comment by jaeilepp** [comment]
Yeah, I can take a look. Where can I find the data?

**Comment by agramfort** [comment]
I just shared the data with you.

**Comment by Eric89GXL** [comment]
Closing in favor of #3915
